zproject/backends.py

```diff
@@ -312,95 +308,80 @@ class EmailAuthBackend(ZulipAuthMixin):
     """
 
     def authenticate(self, username: Optional[str]=None, password: Optional[str]=None,
-                     realm_subdomain: Optional[str]=None,
+                     realm: Optional[Realm]=None,
                      return_data: Optional[Dict[str, Any]]=None) -> Optional[UserProfile]:
         """ Authenticate a user based on email address as the user name. """
         if username is None or password is None:
-            # Return immediately.  Otherwise we will look for a SQL row with
-            # NULL username.  While that's probably harmless, it's needless
-            # exposure.
-            return None
-
-        user_profile = common_get_active_user_by_email(username, return_data=return_data)
-        if user_profile is None:
+            # Because of how we structure our auth calls to always
+            # specify which backend to use when not using
+            # EmailAuthBackend, username and password should always be set.
```

Comment:
The key thing here is that we eliminated the last other backend that used a positional argument for the first parameter to mean something totally different.  And yeah, the approach for that was the `dev_auth_username` thing.

(Because Django calls all backends with a matching argument list, positional arguments with different names all "match" each other)

Reply:
Hmm, I don't quite follow -- in master, our `authenticate` call in `dev_direct_login` looks like this:
```
    user_profile = authenticate(username=email, realm_subdomain=get_subdomain(request))
```
So that looks like not positional arguments to me. In fact all our non-test invocations of `authenticate` pass no positional arguments, if my quick grep just now is accurate.

Sadly the Django docs (e.g. [this page](https://docs.djangoproject.com/en/1.11/topics/auth/customizing/)) aren't super clear on how this works. I might go read the source (though it shouldn't be essential for this branch.)

Reply:
Yeah, the point is that because we don't pass a positional argument that can be treated as the `password` second argument to EmailAuthBackend, this backend doesn't get run.  

We had mostly purged bad behaving `authenticate` calls a couple months ago; but I think the code that make this lose its test coverage was cleaning up one of the other backends that we adjust earlier in this branch (i.e., previously, the unit test for RemoteAuthBackend or whichever it was covered this line).

Reply:
If not otherwise taken care of when we merge this branch, I'd like to hang onto this as a followup TODO: it'd be good for me and other contributors to understand how this `authenticate` dispatch works, and it seems like there's stuff already in your head about it that would be fruitful to write down to that end.

Reply:
Our last two comments raced -- thanks for that further explanation! (Glad I went and skimmed through the email notification thread before archiving it, rather than depend on GitHub's UI.) I think my bottom line is the same, though, as I still don't quite follow your explanation.

We rarely pass any positional arguments to `authenticate` in general, I think -- I see a couple of cases with one positional argument, the request, and a few other cases in tests.

I still don't think I understand how the dispatch of figuring out which backends to call works. Are you able to concisely state your understanding of what that semantics is supposed to be? Or should I just go study the source myself?

Reply:
It's pretty complicated to explain exactly what it does; it's a Django thing designed to help make it easy to just call `authenticate` and pass in some arguments and have it do the right thing.  You can read the code here:

`/srv/zulip-py3-venv/lib/python3.5/site-packages/django/contrib/auth/__init__.py`